apollo-router/src/spec/query.rs

```diff
@@ -239,31 +229,31 @@ impl Query {
                 }
             }
             Some(Value::Null) => {
-                // Detect if root __typename is asked in the original query (the qp doesn't put root __typename in subselections)
-                // cf https://github.com/apollographql/router/issues/1677
-                let operation_kind_if_root_typename = original_operation.and_then(|op| {
-                    op.selection_set
-                        .iter()
-                        .any(|f| f.is_typename_field())
-                        .then(|| *op.kind())
-                });
-                response.data = match operation_kind_if_root_typename {
-                    Some(operation_kind) => {
+                if let Some(operation) = original_operation {
+                    // Detect if root __typename is asked in the original query (the qp doesn't put root __typename in subselections)
+                    // cf https://github.com/apollographql/router/issues/1677
+                    if self.has_only_typename_field(&operation.selection_set, &variables) {
+                        let operation_type_name = schema
+                            .root_operation(operation.kind.into())
+                            .map(|name| name.as_str())
+                            .unwrap_or(operation.kind.default_type_name());
+
                         let mut output = Object::default();
-                        output.insert(TYPENAME, operation_kind.default_type_name().into());
-                        Some(output.into())
+                        output.insert(TYPENAME, operation_type_name.into());
+                        response.data = Some(output.into());
+                        return vec![];
                     }
-                    None => Some(Value::default()),
-                };
+                }
 
+                response.data = Some(Value::Null);
                 return vec![];
             }
             _ => {
                 failfast_debug!("invalid type for data in response. data: {:#?}", data);
             }
         }
 
-        response.data = Some(Value::default());
```

Comment:
I changed `Value::default` to explicit null since it easier to compare with GraphQL spec algorithms